packages/flutter_tools/test/integration.shard/test_data/migrate_project.dart

```diff
@@ -0,0 +1,227 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@Timeout(Duration(seconds: 600))
+
+import 'package:file/file.dart';
+import 'package:flutter_tools/src/globals.dart' as globals;
+
+import '../../src/common.dart';
+import '../test_utils.dart';
+import 'project.dart';
+
+class MigrateProject extends Project {
+  MigrateProject(this.version, {this.vanilla = true});
+
+  @override
+  Future<void> setUpIn(Directory dir, {
+    bool useDeferredLoading = false,
+    bool useSyntheticPackage = false,
+  }) async {
+    this.dir = dir;
+    _appPath = dir.path;
+    if (androidLocalProperties != null) {
+      writeFile(fileSystem.path.join(dir.path, 'android', 'local.properties'), androidLocalProperties);
+    }
+    final Directory tempDir = createResolvedTempDirectorySync('cipd_dest.');
+    final Directory depotToolsDir = createResolvedTempDirectorySync('depot_tools.');
+
+    await processManager.run(<String>[
+      'git',
+      'clone',
+      'https://chromium.googlesource.com/chromium/tools/depot_tools',
+      depotToolsDir.path,
+    ], workingDirectory: dir.path);
+
+    final File cipdFile = depotToolsDir.childFile(globals.platform.isWindows ? 'cipd.bat' : 'cipd');
+    await processManager.run(<String>[
+      cipdFile.path,
+      'init',
+      tempDir.path,
+      '-force',
+    ], workingDirectory: dir.path);
+
+    await processManager.run(<String>[
+      cipdFile.path,
+      'install',
+      'flutter/test/full_app_fixtures/vanilla',
+      version,
+      '-root',
+      tempDir.path,
+    ], workingDirectory: dir.path);
+
+    // This cp command changes the symlinks to real files so the tool can edit them.
+    if (globals.platform.isWindows) {
+      await processManager.run(<String>[
+        'robocopy',
+        tempDir.path,
+        dir.path,
+        '*',
+        '/E',
+        '/mov',
+      ]);
+    } else {
+      await processManager.run(<String>[
+        'cp',
+        '-R',
+        '-L',
+        '-f',
+        '${tempDir.path}/.',
+        dir.path,
+      ]);
+
+      await processManager.run(<String>[
+        'rm',
+        '-rf',
+        '.cipd',
+      ], workingDirectory: dir.path);
+
+      final List<FileSystemEntity> allFiles = dir.listSync(recursive: true);
+      for (final FileSystemEntity file in allFiles) {
+        if (file is! File) {
+          continue;
+        }
+        await processManager.run(<String>[
+          'chmod',
+          '+w',
+          file.path,
+        ], workingDirectory: dir.path);
+      }
+    }
+
+    if (!vanilla) {
+      writeFile(fileSystem.path.join(dir.path, 'lib', 'main.dart'), libMain);
+      writeFile(fileSystem.path.join(dir.path, 'lib', 'other.dart'), libOther);
+      writeFile(fileSystem.path.join(dir.path, 'pubspec.yaml'), pubspecCustom);
+    }
+    tryToDelete(tempDir);
+    tryToDelete(depotToolsDir);
+  }
+
+  final String version;
+  final bool vanilla;
+  late String _appPath;
```

Comment:
could this be set in the constructor?

Reply:
Not in the way the current project class is designed. The directory is only known once we call setUpIn(), which must be called first before any of the other getters/methods are valid. We track the appPath here because some of the getters are dynamically filled with the CIPD downloaded contents.